core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -220,62 +218,65 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     const Kokkos::Experimental::SYCL& space = m_policy.space();
     Kokkos::Experimental::Impl::SYCLInternal& instance =
         *space.impl_internal_space_instance();
-    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMemory& kernelMem =
-        *instance.m_indirectKernel;
-
-    // Allocate USM shared memory for the functor
-    kernelMem.resize(std::max(kernelMem.size(), sizeof(functor)));
+    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMem&
+        indirectKernelMem = instance.m_indirectKernelMem;
 
-    // Placement new a copy of functor into USM shared memory
-    //
-    // Store it in a unique_ptr to call its destructor on scope exit
-    std::unique_ptr<Functor, Kokkos::Impl::destruct_delete> kernelFunctorPtr(
-        new (kernelMem.data()) Functor(functor));
+    // Store a copy of the functor in indirectKernelMem
+    auto kernelFunctorPtr = indirectKernelMem.copy_from(functor);
 
     auto kernelFunctor = ExtendedReferenceWrapper<Functor>(*kernelFunctorPtr);
     sycl_direct_launch(m_policy, kernelFunctor);
   }
 
+  void zero_length_reduction() const {
+    // Convenience references
```

Comment:
FYI, in #3619 it turned out that we really need to run `FunctorFinal` on the device and this requires the functor to possibly be launched indirectly. Hence, I am incorporating the zero-length case into `indirect_launch` again in that pull request.

Reply:
We split the pull request. It's now #3625.